docs/design-proposals/node/Partial-Runtime-Readiness-Handling.md

```diff
@@ -0,0 +1,121 @@
+1. Runtime service new field: isPrimary
+
+        type runtimeService struct {
+            name         string
+            workloadType string
+            endpointUrl  string
+            serviceApi   internalapi.RuntimeService
+            isDefault    bool
+            // primary runtime service is the runtime for Kubernetes system, Daemonset workload containers
+            // primary runtime service must be ready for node readiness condition
+            isPrimary    bool
+        }
+    
+2. Expand NodeCondition array with NodeContainerRuntimeReady and NodeVmRuntimeReady conditions   
+   
+       // registered runtime readiness for each workload type
+        NodeContainerRuntimeReady NodeConditionType = "ContainerRuntimeReady"
+        NodeVmRuntimeReady        NodeConditionType = "VmRuntimeReady"
+
+### Component changes
+
+### Kubelet
+1. Define isPrimary
+
+    Parser change to get the isPrimary field from Kubelet command line parameters. If it is not specified by the cluster 
+    admins, set default to Container runtime.
+    UpdateRuntimeUp() function with new NodeReady runtime criteria with primary runtime service ready.
+    
+3. NodeStatus setter for runtime services
+
+        // RuntimeServiceCondition returns a Setter that update the v1.RuntimeServiceState condition array on the node
+        // the conditions defined in v1.types as:
+        // ContainerRuntimeReady NodeConditionType = "Ready"
+        // VMRuntimeReady NodeConditionType = "Ready"
+        func RuntimeServiceCondition(nowFunc func() time.Time, // typically Kubelet.clock.Now
+        	runtimeServiceStateFunc func() (map[string]map[string]bool, error),
+        	recordEventFunc func(eventType, event string), // typically Kubelet.recordNodeStatusEvent
+        ) Setter {
+            // Setter logic
+        }
+
+
+### Scheduler
+A new predicate function, as well as code to enable the new predicate function.
+
+    // CheckNodeRuntimeReadiness checks if the desired runtime service is ready on a node
+    // Return ture IIF the desired node condition exists, AND the condition is TRUE
+    func CheckNodeRuntimeReadinessPredicate(pod *v1.Pod, meta PredicateMetadata, nodeInfo *schedulernodeinfo.NodeInfo) (bool, []PredicateFailureReason, error) {
+    	// Predicate code
+    }
+
+
+## Discussion
+#### NodeSelectorPredicate vs. NodeConditionCheckPredicate
+Even node selector is commonly used to predicate pod-node binding for some special situations, such as current virtlet 
+Daemonset deployment uses extraRuntime=virtlet as the node selector to ensure the Virtlet runtime service is deployed 
+only on the node with the label. However, given the nature and the current code flow on node readiness with its runtime 
+service state, it is favored to expand the node condition check predicate function as the scheduler for this purpose. 
+so that the node status update routine won't need to update the node labels when runtime state changes, and the POD 
```

Comment:
You don't need to update the node labels, but you need to update NodeStatus, right? :)

Reply:
either update nodestatus( node conditions) or labels will work functionality wise. and they will need to be updated anytime the runtime service state is changed at the node, either labels or node conditions. it is currently decided to use node status simply follow the current code flow for single runtime service status. i am actually open for changes especially the node conditions probably will be "deprecated" in the future.
i will update the doc to make it clear as a least effort to clarify the reasoning here.